File: ts/src/core/regime.ts
/**
 * regime.ts — the NEUTRAL/PRESSURE decision, keyed entirely off ready_tick (spec §2.1) [L2].
 *
 * This is the consistency keystone: the SAME timeline produces both the neutral mind-read and
 * offense/pressure with no special-casing. The engine always asks the entity with the lower
 * ready_tick to choose next; ties mean both are free → the simultaneous hidden commit.
 */
import { type Tick } from "./tick";
import { type Entity } from "./entity";

export type Regime =
  /** Both entities are actionable at the same tick → both commit simultaneously and hidden. */
  | { readonly kind: "NEUTRAL" }
  /** One entity is free while the other is locked → the actor chooses with full information. */
  | { readonly kind: "PRESSURE"; readonly actor: 0 | 1 };

/** Decide the regime from the two entities' ready_ticks (spec §2.1). */
export function computeRegime(a: Entity, b: Entity): Regime {
  if (a.readyTick === b.readyTick) return { kind: "NEUTRAL" };
  return { kind: "PRESSURE", actor: a.readyTick < b.readyTick ? 0 : 1 };
}

/** The next tick at which some entity becomes actionable (the next decision point). */
export function nextDecisionTick(a: Entity, b: Entity): Tick {
  return a.readyTick < b.readyTick ? a.readyTick : b.readyTick;
}

File: rust/engine/src/fighting/regime.rs
//! The NEUTRAL vs PRESSURE decision regime — derived *entirely* from `ready_tick`, with no
//! special-casing (spec §2.1, mechanics §1.3). The concrete derivation lives in [`super::sim`]
//! (`pending_decision`); this is just the label the engine reports.

/// Which decision regime the fight is in.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Regime {
    /// Every contender is free at once → commit hidden, simultaneously (the neutral mind-read).
    Neutral,
    /// One entity is free while ≥1 opponent is locked → it chooses with full information.
    Pressure,
}
